lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp

```diff
@@ -603,34 +603,35 @@ void mlir::torch::onnx_c::populateDefaultDomainAtoF(
                 });
   patterns.onOp(
       "Clip", 13, [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        // Note: Clip existed in older versions but with some differences:
+        // https://onnx.ai/onnx/operators/onnx__Clip.html. Notably:
+        //   * Prior to version 12, only floating point values were accepted.
+        //   * In version 6, the Clip op used 1 input with attributes for
+        //     min/max instead of 1-3 inputs.
+
+        // Inputs and outputs must be tensors. Min and max are both optional
+        // and if provided must be "tensors of empty shape" (scalars).
+        Value source;
         Torch::ValueTensorType resultType;
+        if (binder.tensorOperandAtIndex(source, 0) ||
+            binder.tensorResultType(resultType)) {
+          return failure();
+        }
+
```

Comment:
I would restructure this to avoid having exit points decided by `getNumOperands`. We know that there are versions of `clip` that include attribute defined values. Do something like the following

```
Value min, max;
if (binder.op->getNumOperands() > 1) {
  if (binder.tensorOperandAtIndex(min, 1)
    return rewriter.notifyMatchFailure(binder.op, "error message");
}

if (binder.op->getNumOperands() > 2) {
  if (binder.tensorOperandAtIndex(max, 2)
    return rewriter.notifyMatchFailure(binder.op, "error message");
}

// Can insert attribute based checking with value formation here.

if (!min && !max) {
  rewriter.replaceOp(binder.op, operand);
}

// Min case

// Max case

// Clamp case
```

Reply:
>  We know that there are versions of clip that include attribute defined values.

Would you like those to be supported? That would be version 6, when this op lowering is currently using a minimum version of 13. I figure the ONNX Version Converter (https://github.com/onnx/onnx/blob/main/docs/VersionConverter.md) should be used to support older versions, rather than include potentially complicated switching logic both here _and_ in that upstream tool.

Reply:
> ```c++
> if (binder.op->getNumOperands() > 1) {
>   if (binder.tensorOperandAtIndex(min, 1)
>     return rewriter.notifyMatchFailure(binder.op, "error message");
> }
> ```
> 

These also have the same bug as before - the operands can be 'none', which is not a tensor type (? I'm not sure exactly what type, should run under a debugger, but the `tensorOperandAtIndex` check was failing)

Reply:
Tried restructuring how you suggested but it got messy and loose with validation (https://github.com/ScottTodd/torch-mlir/commit/b6c307b023b52adc3a46e09c33c16298fc6d52e9).

Instead, I added a second pattern, like what is done here: https://github.com/llvm/torch-mlir/blob/933db87a07fb828b17b75f2f5f396a434f8f1a17/lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp#L1038-L1040

How does this look now?

Reply:
Try to embed similar to ReduceMin. The omni-rewriter is a weird restriction we run in and recursion limits applying a single monolithic rewrite pattern could cause problems.

https://github.com/llvm/torch-mlir/blob/main/lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp#L1178

Reply:
Refactored a bit. LMK what you think... seems to be getting better.